### PR TITLE
Pull in class-based genreconciler

### PR DESCRIPTION
**Pull request by mattmoor**
WIP until https://github.com/knative/networking/pull/41 lands.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -30,4 +30,7 @@ replace (
 	k8s.io/apimachinery => k8s.io/apimachinery v0.17.6
 	k8s.io/client-go => k8s.io/client-go v0.17.6
 	k8s.io/code-generator => k8s.io/code-generator v0.17.6
+
+	// TODO(mattmoor): DO NOT SUBMIT
```
Comment (mattmoor) [inline go.mod:34]: Found "DO NOT SUBMIT".

**Comment by mattmoor** [comment]
This test seems to be failing consistently post-submit: https://testgrid.knative.dev/net-kourier#continuous&exclude-non-failed-tests=50&sort-by-flakiness=50&width=20

**Comment by mattmoor** [comment]
I rebased this on the real change, but I'm guessing the HA failing in CI is going to block this.

cc @nak3 @markusthoemmes

**Comment by nak3** [comment]
LGTM.

Yes, we need to fix the HA failing in CI :sweat_drops:

**Comment by jmprusi** [comment]
@nak3 I'm working on it... let's see If I can find the solution today

**Pull request merged**